tiatoolbox/tools/registration/wsi_registration.py

```diff
@@ -172,3 +176,259 @@ def match_histograms(image_a, image_b, kernel_size=7):
+
+        Args:
+            feature_dist (:class:`numpy.ndarray`):
+                A feature distance array.
+
+        Returns:
+            :class:`numpy.ndarray`:
+                An array of matching points.
+            :class:`numpy.ndarray`:
+                An array of floating numbers representing quality
+                of each matching points.
+
+        """
+        seq = np.arange(feature_dist.shape[0])
+        ind_first_min = np.argmin(feature_dist, axis=1)
+        first_min = feature_dist[seq, ind_first_min]
+        mask = np.zeros_like(feature_dist)
+        mask[seq, ind_first_min] = 1
+        masked = np.ma.masked_array(feature_dist, mask)
+        second_min = np.amin(masked, axis=1)
+        return np.array([seq, ind_first_min]).transpose(), np.array(
+            second_min / first_min
+        )
+
+    @staticmethod
+    def compute_feature_distance(feature_x, feature_y, factor):
+        """Computes feature distance.
+
+        This function computes Euclidean distance between features of
+        fixed and moving images.
+
+        Args:
+            feature_x (:class:`numpy.ndarray`):
+                Features computed for a fixed image.
+            feature_y (:class:`numpy.ndarray`):
+                Features computed for a moving image.
+            factor (int):
+                A number multiplied by the feature size
+                for getting the referenced feature size.
+
+        Returns:
+            :class:`numpy.ndarray`:
+                A feature distance array.
+
+        """
+        feature_distance = np.linalg.norm(
+            np.repeat(np.expand_dims(feature_x, axis=0), feature_y.shape[0], axis=0)
+            - np.repeat(np.expand_dims(feature_y, axis=1), feature_x.shape[0], axis=1),
+            axis=len(feature_x.shape),
+        )
+
+        feature_size_2d = np.int(np.sqrt(feature_distance.shape[0]))
+        ref_feature_size_2d = factor * feature_size_2d
+        feature_size, ref_feature_size = feature_size_2d**2, ref_feature_size_2d**2
+        feature_grid = np.kron(
+            np.arange(feature_size).reshape([feature_size_2d, feature_size_2d]),
+            np.ones([factor, factor], dtype="int32"),
+        )
+        row_ind = np.repeat(
+            feature_grid.reshape([ref_feature_size, 1]), ref_feature_size, axis=1
+        )
+        col_ind = np.repeat(
+            feature_grid.reshape([1, ref_feature_size]), ref_feature_size, axis=0
+        )
+        return feature_distance[row_ind, col_ind]
+
+    def feature_mapping(self, features, num_matching_points=128):
+        """Mapping of CNN features.
+
+        This function maps features of a fixed image to that of
+        a moving image on the basis of Euclidean distance between
+        them.
+
+        Args:
+            features (dict):
+                Multiscale CNN features.
+            num_matching_points (int):
+                Number of required matching points.
+
+        Returns:
```

Comment:
See previous comment about tuple return types in docstrings

Reply:
https://github.com/TissueImageAnalytics/tiatoolbox/blob/f701395bd9ff7d6c72c6a2db710a945d638cbccc/tiatoolbox/tools/registration/wsi_registration.py#L438-L446